lms/djangoapps/program_enrollments/tests/test_signals.py

```diff
@@ -0,0 +1,83 @@
+"""
+Test signal handlers for program_enrollments
+"""
+
+from __future__ import absolute_import
+import mock
+
+from lms.djangoapps.program_enrollments.api.v1.tests.factories import ProgramEnrollmentFactory
+from lms.djangoapps.program_enrollments.models import ProgramEnrollment
+from lms.djangoapps.program_enrollments.signals import _listen_for_lms_retire
+from openedx.core.djangoapps.user_api.accounts.tests.retirement_helpers import fake_completed_retirement
+from student.tests.factories import UserFactory
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+
+
+class SurveyRetireSignalTests(ModuleStoreTestCase):
+    """
+    Test the _listen_for_lms_retire signal
+    """
+
+    def create_enrollment_and_history(self, user=None):
+        """
+        Create ProgramEnrollment and several History entries
+        """
+        if user:
+            enrollment = ProgramEnrollmentFactory(user=user)
+        else:
+            enrollment = ProgramEnrollmentFactory()
+        for status in ['pending', 'suspended', 'withdrawn', 'enrolled']:
+            enrollment.status = status
+            enrollment.save()
+        return enrollment
+
+    def assert_enrollment_and_history_retired(self, enrollment):
+        """
+        Assert that for the enrollment and all histories, external key is None
+        """
+        enrollment.refresh_from_db()
+        self.assertIsNone(enrollment.external_user_key)
+        for history_record in enrollment.historical_records.all():
+            self.assertIsNone(history_record.external_user_key)
+
+    def test_retire_enrollment(self):
+        """
+        Test basic retirement of program enrollment
+        """
+        enrollment = self.create_enrollment_and_history()
+        _listen_for_lms_retire(sender=self.__class__, user=enrollment.user)
+        self.assert_enrollment_and_history_retired(enrollment)
+
+    def test_retire_enrollment_multiple(self):
+        """
+        Test basic retirement of user with multiple program enrollments
+        """
+        enrollment = self.create_enrollment_and_history()
+        enrollment2 = self.create_enrollment_and_history(user=enrollment.user)
+        enrollment3 = self.create_enrollment_and_history(user=enrollment.user)
+        _listen_for_lms_retire(sender=self.__class__, user=enrollment.user)
+        self.assert_enrollment_and_history_retired(enrollment)
+        self.assert_enrollment_and_history_retired(enrollment2)
+        self.assert_enrollment_and_history_retired(enrollment3)
+
+    @mock.patch.object(ProgramEnrollment, 'retire_user')
+    def test_success_no_enrollment(self, mocked_retire):
+        """
+        Basic success path for users who have no enrollments, should simply not error
+        """
+        user = UserFactory()
+        _listen_for_lms_retire(sender=self.__class__, user=user)
+        mocked_retire.assert_not_called()
```

Comment:
This sadly doesn't do what you want in python 2.7 - `assert_not_called` wasn't added to the mock library until version 3.5 of python (in python 2.7, mock is an external library, and not part of the standard lib).
```
alexdusenbery@aed-mbp:devstack (master) $ make lms-shell
docker exec -it edx.devstack.lms env TERM=xterm-256color /edx/app/edxapp/devstack.sh open
root@lms:/edx/app/edxapp/edx-platform# ./manage.py lms shell
...
>>> import mock
>>> m = mock.Mock()
>>> dir(m)
['assert_any_call', 'assert_called_once_with', 'assert_called_with', 'assert_has_calls', 'attach_mock', 'call_args', 'call_args_list', 'call_count', 'called', 'configure_mock', 'method_calls', 'mock_add_spec', 'mock_calls', 'reset_mock', 'return_value', 'side_effect']
```
The reason this statement doesn't result in an exception is because mock objects have all attributes that are invoked on them, i.e. `mocked_retire.foobar()` is a completely valid (but useless) statement.
What you want to use is `self.assertFalse(mocked_retire.called)`.

Reply:
>:[ python twooooooo \*shakes fist at sky\*

Reply:
in that same vein, I recommend putting `autospec=True` on the `mock.patch.object` call.

Reply:
this was an incorrect assertion anyway. retire_user is still gonna be called, it's just not gonna do anything